feat: AE excretion calculations with PKNCA function

js3110 opened the pull request:
## Issue

Closes #429 #470 #452 

## Description

Calculation of PKNCA excretion parameters:

Including:
- Amount excreted (pk.calc.ae): "RCAMINT" , "Amt Rec from T1 to T2" (Ae)
-  ~Fraction excreted (pk.calc.fe): "RCPCINT", "Pct Rec from T1 to T2" (Fe)~ (removed since PKNCA does not convert units. This will be updated in the next PKNCA push to CRAN)

## Definition of Done

The requirements for the feature to be complete:
- [x] Dataset including urine data can be uploaded and run through NCA with no issues
- [x] Calculation of Urine AE, ~fraction excreted~.
- [x] Results shown in NCA table
- [x] Results in Non-NCA data in Excretion Analysis section

## How to test

Use `adnca_dummy_sm_dataset.csv` from tests/testthat/data folder. Try to add urine parameters to main NCA, and also go to Additional Analysis -> Excretion.  Test with Urine or Feces PCSPEC.

## Contributor checklist
- [x] Code passes lintr checks
- [x] Code passes all unit tests
- [x] New logic covered by unit tests
- [x] New logic is documented
- [x] Package version is incremented

## Notes to reviewer

There are two workflows added, one for the main NCA and one for the excretion analysis tab. The excretion analysis tab is the main part, with NCA results working but likely not to be used.
I have integrated part of #436 here, by preventing excretion parameters from being calculated on non-excretion samples, and preventing non excretion parameters being calculated on excretion samples.
**Note**: currently FREXINT, and RENALCL do not work within PKNCA- they can be selected as parameters but the calculations are not possible/correct yet. This is something that Bill is working on in the PKNCA package.

@rabiibouhestine I added a styling section for the excretion tab, I tried to do it so that it aligns with the rest of the styling, to make it easier to update later. Could you take a look at this part and let me know if its all ok?

js3110 commented:
@Shaakon35 I have updated so urine parameters are automatically mapped, and the VOLUME and VOLUMEU lables are included in the mapping.  As discussed the issue with the dataset you tried is a data issue, not an aNCA issue.
PTAL :)

m-kolomanski reviewed (changes requested):
Hey, I have some comments for the PR.

This PR is quite sizeable, for the future please do try to split PRs into smaller chunks if there is potential. Here, I could see the package functions for conversions and formatting being a separate PR, and then having a second submission for app implementation with excretion_analysis module - this would make the PR more compact and the review easier and faster.

Changes to the logic in the `R/` package do not have to immediately come with application implementation and could be submitted as separate PRs, within the scope of the same task.

js3110 commented on R/conversions.R:
```diff
@@ -84,3 +84,129 @@ convert_to_iso8601_duration <- Vectorize(function(value, unit) {
+                                 volumeu = "VOLUMEU") {
+
+  required_cols <- c(avalu, volume, volumeu)
+  if (!all(required_cols %in% names(df))) {
```
> question: Should we throw at least a warning in that case?

yes good point, i would say in the logs but not on the UI?

js3110 commented on inst/shiny/modules/tab_nca/excretion_analysis.R:
```diff
@@ -0,0 +1,207 @@
+      conc_data <- input_pknca_data()$conc$data
+      volume_missing <- !all(c("VOLUME", "VOLUMEU") %in% names(conc_data))
+
+      if (volume_missing) {
+        session$sendCustomMessage("showOverlay", list(id = session$ns("excretion_overlay")))
+      } else {
+        session$sendCustomMessage("hideOverlay", list(id = session$ns("excretion_overlay")))
+      }
```
will this still maintain the formatting of the hidden panel ? (with the message advising the user to map volume column?)

js3110 commented:
@m-kolomanski thanks for the review! I agree its a bit too long, so I am sorry about that! I tried to start small but got a bit carried away...

I have fixed everything from your comments, just asked a couple questions along the way for my knowledge.

I've also updated the logs for the volume conversion, let me know if you think its a bit overkill or not (imo its good, because its important to inform the user of any changes to the data that happen in the backend)

m-kolomanski commented on inst/shiny/modules/tab_nca/excretion_analysis.R:
```diff
@@ -0,0 +1,196 @@
+    })
+
+    # Perform calculations
+    analysis_result <- eventReactive(input$submit_btn, {
```
issue(non-blocking): `reactive() |> bindEvent()` is currently the preferred solution.

m-kolomanski reviewed (approved):
LGTM, thank you for the changes! I have two non-blocking changes, otherwise looks good.

Gero1999 commented on tests/testthat/test-format_data.R:
```diff
@@ -342,4 +342,49 @@ describe("format_pkncadata_intervals", {
+    expect_true(all(result_volume$ae))
+    expect_true(all(result_volume$fe))
+  })
+
+  it("sets correct parameters if VOLUME not present", {
+    pknca_conc_no_volume <- PKNCA::PKNCAconc(
+      df_conc,
+      formula = AVAL ~ TIME | STUDYID + PCSPEC + DRUG + USUBJID / PARAM,
+      exclude_half.life = "exclude_half.life",
+      time.nominal = "NFRLT"
+    )
+
+    result_no_volume <- format_pkncadata_intervals(
+      pknca_conc_no_volume,
+      pknca_dose, params = c("ae", "fe", "cmax", "tmax", "half.life")
+    )
+
+    expect_false(all(result_no_volume$ae))
+    expect_false(all(result_no_volume$fe))
+    expect_true(all(result_no_volume$cmax))
+    expect_true(all(result_no_volume$tmax))
+    expect_true(all(result_no_volume$half.life))
+  })
+
```
issue: redundant test
```suggestion

```

Gero1999 replied:
the logic is already covered by all the previous ones, is not doing nothing special. The "exception" is having the volume mapped

js3110 replied:
I disagree, this test is specficially checking that "ae"and "fe" are set to false if volume is not present. There are no other tests that do that currently.

Gero1999 replied:
you are right, let's keep it

Gero1999 commented on tests/testthat/test-format_data.R:
```diff
@@ -342,4 +342,49 @@ describe("format_pkncadata_intervals", {
+      params = c("ae", "fe", "cmax", "tmax", "half.life", "cl.obs")
+    )
+
+    expect_false(all(result_volume$cmax))
+    expect_false(all(result_volume$tmax))
+    expect_false(all(result_volume$half.life))
+    expect_false(all(result_volume$cl.obs))
```
issue: we should not need to evaluate these parameters. but lmk if I am missing something!

```suggestion
```

Gero1999 replied:
hmm I see, so if the user chooses `cmax` you won't allow them to calculate it for urine samples because you assume they don't want it... And you will do the same with any parameter that is not `ae`, `fe`. Am I right? 

If so I see 3 problems with this approach:

- `Hidden action for the user`. I am concerned if this is not intuitive, I prefer the risk of the user to have more parameters than less of what they want.

- `Parameter ratios`. We want PKNCA to be able to produce at least: `cmax`, `aucxxx`

- `Documentation`. The function `verify_parameters()` clearly says that `ae` & `fe` are added, but I don't remember it saying nothing about removing all other parameters.

I would personally opt for `verify_parameters` to keep all parameters and just add on top `ae`, `fe`. This would be cleaner and prevent new bugs/issues for future implementations

Gero1999 replied:
yes please remove the changes to the intervals, just add `ae`, `fe`. The solution for #436 has to be different

Gero1999 commented on tests/testthat/test-format_data.R:
```diff
@@ -342,4 +342,49 @@ describe("format_pkncadata_intervals", {
+    expect_false(all(result_volume$tmax))
+    expect_false(all(result_volume$half.life))
+    expect_false(all(result_volume$cl.obs))
+    expect_true(all(result_volume$ae))
+    expect_true(all(result_volume$fe))
```
question: here you are checking the intervals. I am assuming that afterwards calculations are done using exclusively `pk.nca` and there is no post-processing for `ae`, `fe` right?

js3110 replied:
yep, exactly

Gero1999 reviewed (changes requested):
lgtm, just some small changes to look at

js3110 commented on R/PKNCA.R:
```diff
@@ -95,27 +102,47 @@ PKNCA_create_data_object <- function(adnca_data) { # nolint: object_name_linter
-    exclude_half.life = "exclude_half.life",
-    include_half.life = "include_half.life",
-    time.nominal = "NFRLT",
-    concu = "AVALU",
-    timeu = "RRLTU"
+  # Create PKNCA conc object
+  if (volume_column %in% colnames(df_conc)) {
+
+    # ensure units are correct for excretion calculations
+    df_conc <- convert_volume_units(df_conc)
+
+    pknca_conc <- PKNCA::PKNCAconc(
+      df_conc,
+      formula = AVAL ~ TIME | STUDYID + PCSPEC + DRUG + USUBJID / PARAM,
+      exclude_half.life = "exclude_half.life",
+      include_half.life = "include_half.life",
+      time.nominal = "NFRLT",
+      volume = volume_column,
+      concu = "AVALU",
+      timeu = "RRLTU",
+      amountu = "AMOUNTU"
+    )
+
+  } else {
+    # Create PKNCA objects
+    pknca_conc <- PKNCA::PKNCAconc(
+      df_conc,
+      formula = AVAL ~ TIME | STUDYID + PCSPEC + DRUG + USUBJID / PARAM,
+      exclude_half.life = "exclude_half.life",
+      include_half.life = "include_half.life",
+      time.nominal = "NFRLT",
+      concu = "AVALU",
+      timeu = "RRLTU"
+    )
+  }
+
+  # Create dosing data
+  df_dose <- format_pkncadose_data(
+    pkncaconc_data = df_conc,
+    group_columns = c(group_columns, usubjid_column)
```
a lot of the later analysis relies on determining whether volume is mapped or not, so im hesitant to automatically map volume and create amountu if its not needed. I would prefer to keep two separate pathways

Gero1999 reviewed (approved):
hey @js3110 before merging regarding this when I test the `adnca_dummy_sm_dataset.csv` with default parameters selected I only see in the results the two excretion parameters. (no `cmax`...). Is this only conditioned for the displayed results or also for the `res_nca` object?
